runners/spark/3/src/main/java/org/apache/beam/runners/spark/structuredstreaming/translation/PipelineTranslator.java

```diff
@@ -80,6 +83,12 @@
 public abstract class PipelineTranslator {
   private static final Logger LOG = LoggerFactory.getLogger(PipelineTranslator.class);
 
+  // Threshold to limit query plan complexity to avoid unnecessary planning overhead. Currently this
+  // is fairly low, Catalyst won't be able to optimize beyond ParDos anyways. Until there's
+  // dedicated support for schema transforms, there's little value of allowing more complex plans at
+  // this point.
+  private static final int PLAN_COMPLEXITY_THRESHOLD = 6;
```

Comment:
Why it is 6? Should it be configurable?

Reply:
That was a decent value when testing. There's no point in having this configurable, that way users would only worry about it.

Reply:
Honestly, this is mostly optimizing for the benchmark ... the proportional overhead of the planner will decrease the more data there is.

Reply:
Can this value affect a pipeline performance significantly depending on user case (like a complexity of pipeline, size and type of input data, etc)?

Reply:
No, as said, it only optimizes the Calcite planner, the size of the input is irrelevant.

Reply:
Yes but planner depends on the complexity of pipeline (like initial cause of this PR and TCPDS query 83). 
I'm ok to keep it hardcoded but I don't see any disadvantages to keep this default value and make it configurable.

Reply:
At this point i strongly think this should not be exposed to the user ... it's deeply internal and more confusing than helpful. Happy to reevaluate if something comes up that proves otherwise.